src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs

```diff
@@ -206,24 +260,105 @@ public static LogicalLocation CreateLogicalLocation(LogicalLocationVersionOne v1
             return logicalLocation;
         }
 
-        public static IList<Notification> CreateNotificationsList(IEnumerable<NotificationVersionOne> v1Notifications)
+        internal LogicalLocation CreateLogicalLocation(string fullyQualifiedLogicalName, string parentKey = null, string decoratedName = null)
         {
-            List<Notification> notifications = null;
+            var logicalLocation = new LogicalLocation
+            {
+                DecoratedName = decoratedName,
+                FullyQualifiedName = fullyQualifiedLogicalName,
+                Name = GetLogicalLocationName(fullyQualifiedLogicalName),
+                ParentKey = parentKey
+            };
+
+            return logicalLocation;
+        }
 
-            if (v1Notifications != null)
+        internal string AddLogicalLocation(LogicalLocation logicalLocation)
+        {
+            if (_currentRun.LogicalLocations == null)
             {
-                notifications = new List<Notification>();
+                _currentRun.LogicalLocations = new Dictionary<string, LogicalLocation>();
+            }
+
+            string fullyQualifiedName = logicalLocation.FullyQualifiedName;
+            string logicalLocationKey = logicalLocation.FullyQualifiedName;
+            int disambiguator = 0;
 
-                foreach (NotificationVersionOne notification in v1Notifications)
+            // We need to do the comparison here too
+            while (_currentRun.LogicalLocations.ContainsKey(logicalLocationKey))
+            {
+                LogicalLocation logLoc = _currentRun.LogicalLocations[logicalLocationKey].DeepClone();
+                logLoc.Name = GetLogicalLocationName(fullyQualifiedName);
+                logLoc.FullyQualifiedName = logicalLocation.FullyQualifiedName;
+
+                if (logicalLocation.ValueEquals(logLoc))
                 {
-                    notifications.Add(CreateNotification(notification));
+                    break;
                 }
+
+                logicalLocationKey = logicalLocation.FullyQualifiedName + "-" + disambiguator.ToString(CultureInfo.InvariantCulture);
+                disambiguator++;
+            }
+
+            if (!_currentRun.LogicalLocations.ContainsKey(logicalLocationKey))
+            {
+                _currentRun.LogicalLocations.Add(logicalLocationKey, logicalLocation);
+                FixUpLogicalLocation(logicalLocationKey);
+            }
+
+            return logicalLocationKey;
+        }
+
+        internal string GetLogicalLocationName(string fullyQualifiedLogicalName)
+        {
+            if (string.IsNullOrWhiteSpace(fullyQualifiedLogicalName))
+            {
+                throw new ArgumentNullException(nameof(fullyQualifiedLogicalName));
+            }
+
+            return fullyQualifiedLogicalName.Split(Utilities.FullyQualifiedNameDelimiters,
+                                                   StringSplitOptions.RemoveEmptyEntries).Last();
+        }
```

Comment:
Again, I feel like I've seen this code before. I think there's some DRYing out to be done between the converters and the transformer. If you agree, let's merge what we have, then do a refactoring PR to clean this up.

Reply:
Sounds good.